Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go

```diff
@@ -0,0 +1,365 @@
+		if ok {
+			close(endpointsAvail)
+			// this shouldn't happen (ok should always be false)
+		}
+	case <-time.NewTimer(needPodsWaitTimeout).C:
+		if allConns.Len() > 0 {
+			glog.Errorf("timed out %v TCP connections while waiting for idled service %s/%s:%s to awaken.", allConns.Len(), service.Namespace, service.Name, service.Port)
+			allConns.Clear()
+		}
+		return
+	}
+	glog.V(4).Infof("unidling TCP proxy got endpoints for service %s/%s:%s, connecting %v accumulated connections", service.Namespace, service.Name, service.Port, allConns.Len())
+
+	for _, inConn := range allConns.GetConns() {
+		glog.V(3).Infof("Accepted TCP connection from %v to %v", inConn.RemoteAddr(), inConn.LocalAddr())
+		outConn, err := tryConnect(service, inConn.(*net.TCPConn).RemoteAddr(), "tcp", proxier)
+		if err != nil {
+			glog.Errorf("Failed to connect to balancer: %v", err)
+			inConn.Close()
+			continue
+		}
+		// Spin up an async copy loop.
+		go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
+	}
+}
+
+// udpUnidlerSocket implements proxySocket.  Close() is implemented by net.UDPConn.  When Close() is called,
+// no new connections are allowed and existing connections are broken.
+// TODO: We could lame-duck this ourselves, if it becomes important.
+type udpUnidlerSocket struct {
+	udpProxySocket
+	eventRecorder record.EventRecorder
+}
+
+func (udp *udpUnidlerSocket) readFromSock(buffer []byte, myInfo *serviceInfo) bool {
+	if !myInfo.isAlive() {
+		// The service port was closed or replaced.
+		return false
+	}
+
+	// Block until data arrives.
+	// TODO: Accumulate a histogram of n or something, to fine tune the buffer size.
+	_, _, err := udp.ReadFrom(buffer)
+	if err != nil {
+		if e, ok := err.(net.Error); ok {
+			if e.Temporary() {
+				glog.V(1).Infof("ReadFrom had a temporary failure: %v", err)
+				return true
+			}
+		}
+		glog.Errorf("ReadFrom failed, exiting ProxyLoop: %v", err)
+		return false
+	}
+
+	return true
+}
+
+func (udp *udpUnidlerSocket) sendWakeup(service proxy.ServicePortName, myInfo *serviceInfo) (chan<- interface{}, *time.Timer) {
+	stopChan := make(chan interface{})
+	timeoutTimer := time.NewTimer(needPodsWaitTimeout)
+
+	go func(stop <-chan interface{}, timeout <-chan time.Time) {
+		select {
+		case <-stop:
+			return
+		case <-timeout:
+			return
+		case <-time.NewTicker(needPodsWaitTimeout).C:
+			glog.V(4).Infof("unidling proxy sent unidle event to wake up service %s/%s:%s", service.Namespace, service.Name, service.Port)
+			udp.eventRecorder.Eventf(myInfo.ref, unidling.NeedPodsReason, "The service-port %s/%s:%s needs pods (at %v)!", service.Namespace, service.Name, service.Port, time.Now())
+		}
+	}(stopChan, timeoutTimer.C)
+
+	return stopChan, timeoutTimer
+}
+
+func (udp *udpUnidlerSocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
+	// just drop the packets on the floor until we have endpoints
+	var buffer [4096]byte // 4KiB should be enough for most whole-packets
```

Comment:
Yeah proxy is not always great code (it's some of the surviving code from the very first kube days - we had to write our own functions in those days, and merge commits uphill both ways in the snow)